Mathlib/CategoryTheory/Localization/SmallHom.lean

```diff
@@ -0,0 +1,273 @@
+/-
+Copyright (c) 2024 Joël Riou. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joël Riou
+-/
+import Mathlib.CategoryTheory.Localization.HomEquiv
+import Mathlib.Logic.Small.Defs
+
+/-!
+# Shrinking morphisms in localized categories
+
+Given a class of morphisms `W : MorphismProperty C`, and two objects `X` and `Y`,
+we introduce a type-class `HasSmallLocalizedHom.{w} W X Y` which expresses
+that in the localized category with respect to `W`, the type of morphisms from `X`
+to `Y` is `w`-small for a certain universe `w`. Under this assumption,
+we define `SmallHom.{w} W X Y : Type w` as the shrunk type. For any localization
+functor `L : C ⥤ D` for `W`, we provide a bijection
+`SmallHom.equiv.{w} W L : SmallHom.{w} W X Y ≃ (L.obj X ⟶ L.obj Y)` that is compatible
+with the composition of morphisms.
+
+-/
+
+universe w w' v₁ v₂ v₃ v₄ v₅ u₁ u₂ u₃ u₄ u₅
+
+namespace CategoryTheory
+
+open Category
+
+namespace Localization
+
+variable {C : Type u₁} [Category.{v₁} C] (W : MorphismProperty C)
+  {D : Type u₂} [Category.{v₂} D]
+  {D' : Type u₃} [Category.{v₃} D']
+
+section
+
+variable (L : C ⥤ D) [L.IsLocalization W] (X Y Z : C)
+
+/-- This property holds if the type of morphisms between `X` and `Y`
+in the localized category with respect to `W : MorphismProperty C`
+is small. -/
+class HasSmallLocalizedHom : Prop where
+  small : Small.{w} (W.Q.obj X ⟶ W.Q.obj Y)
+
+attribute [instance] HasSmallLocalizedHom.small
+
+variable {X Y Z}
+
+lemma hasSmallLocalizedHom_iff :
+    HasSmallLocalizedHom.{w} W X Y ↔ Small.{w} (L.obj X ⟶ L.obj Y) := by
+  constructor
+  · intro h
+    have := h.small
+    exact small_map (homEquiv W W.Q L).symm
+  · intro h
+    exact ⟨small_map (homEquiv W W.Q L)⟩
+
+lemma hasSmallLocalizedHom_of_isLocalization :
+    HasSmallLocalizedHom.{v₂} W X Y := by
+  rw [hasSmallLocalizedHom_iff W L]
+  infer_instance
+
+variable (X Y) in
+lemma small_of_hasSmallLocalizedHom [HasSmallLocalizedHom.{w} W X Y] :
+    Small.{w} (L.obj X ⟶ L.obj Y) := by
+  rwa [← hasSmallLocalizedHom_iff W]
+
+lemma hasSmallLocalizedHom_iff_of_isos {X' Y' : C} (e : X ≅ X') (e' : Y ≅ Y') :
+    HasSmallLocalizedHom.{w} W X Y ↔ HasSmallLocalizedHom.{w} W X' Y' := by
+  simp only [hasSmallLocalizedHom_iff W W.Q]
+  exact small_congr (Iso.homCongr (W.Q.mapIso e) (W.Q.mapIso e'))
+
+end
+
+/-- The type of morphisms from `X` to `Y` in the localized category
+with respect to `W : MorphismProperty C` that is shrunk to `Type w`
+when `HasSmallLocalizedHom.{w} W X Y` holds. -/
+def SmallHom (X Y : C) [HasSmallLocalizedHom.{w} W X Y] : Type w :=
```

Comment:
I'm thinking maybe this should be defined for all `Small (X ⟶ Y)`? And then maybe a thin wrapper (abbrev?) over that to preserve the typeclass inference capabilities we currently have. Not sure how easy the second part is though.

Reply:
No, there must we a specific API for localization, because the morphisms here are computed in a different category from `C` (in happens in the localized category), and the point of all this API is that we can relate these types `SmallHom` (which depends only on `X`, `Y` and the class `W`) by using any functor `L : C ⥤ D` which satisfies the "universal property" of the localization.